Registry layer relies on standard plugin

**Describe the bug**
A core part of the framework (registry layer) relies on pslist to figure out the process that the registry lives in.  This is an awkward introduces a lot of cross-dependency that shouldn't exist.  We should try to find a way to achieve the same goal without pulling in so much extra gubbins just to find the process list.  I'd almost be happy duplicating code, or storing it in a separate common module (perhaps the generic process work from #1000 could help)?

See [#1307](https://github.com/volatilityfoundation/volatility3/pull/1307#discussion_r1817447132) for more details.

Hey @ikelos I'm having this circular import issue with pytest in my local VSCode setup. Since this happens in the pytest test discovery stage, that aborts the execution of all test cases. 

As a temporary workaround, I usually rename the `test_scheduled_tasks.py` file to `test_scheduled_tasks.py_`, since this file doesn’t seem to be used anywhere. However, we should prioritize resolving this issue to restore the development environment to full functionality.

```
==================================== ERRORS ====================================
________ ERROR collecting test/plugins/windows/test_scheduled_tasks.py _________
ImportError while importing test module '/home/gmoreira/vol3/test/plugins/windows/test_scheduled_tasks.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/plugins/windows/test_scheduled_tasks.py:6: in <module>
    from volatility3.plugins.windows import scheduled_tasks
volatility3/framework/plugins/windows/scheduled_tasks.py:18: in <module>
    from volatility3.framework.layers import registry
volatility3/framework/layers/registry.py:17: in <module>
    from volatility3.plugins.windows import pslist
volatility3/framework/plugins/windows/pslist.py:14: in <module>
    from volatility3.framework.symbols.windows.extensions import pe
volatility3/framework/symbols/windows/__init__.py:8: in <module>
    from volatility3.framework.symbols.windows.extensions import pe, pool, registry
volatility3/framework/symbols/windows/extensions/registry.py:11: in <module>
    from volatility3.framework.layers.registry import (
E   ImportError: cannot import name 'RegistryFormatException' from partially initialized module 'volatility3.framework.layers.registry' (most likely due to a circular import) (/home/gmoreira/vol3/volatility3/framework/layers/registry.py)
=========================== short test summary info ============================
ERROR test/plugins/windows/test_scheduled_tasks.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
===================== 54 tests collected, 1 error in 0.28s =====================
```

--- a/volatility3/framework/layers/registry.py
+++ b/volatility3/framework/layers/registry.py
@@ -14,7 +14,7 @@ from volatility3.framework.configuration.requirements import (
 from volatility3.framework.exceptions import InvalidAddressException
 from volatility3.framework.layers import linear
 from volatility3.framework.symbols import intermed
-from volatility3.plugins.windows import pslist
+from volatility3.framework.symbols.windows import extensions
 
 vollog = logging.getLogger(__name__)
 
@@ -65,16 +65,15 @@ class RegistryHive(linear.LinearlyMappedLayer):
 
         # Win10 17063 introduced the Registry process to map most hives.  Check
         # if it exists and update RegistryHive._base_layer
-        for proc in pslist.PsList.list_processes(
-            context=self.context, kernel_module_name=self.config["kernel_module_name"]
-        ):
-            proc_name = proc.ImageFileName.cast(
-                "string", max_length=proc.ImageFileName.vol.count, errors="replace"
+        try:
+            registry_proc = self._find_registry_process()
+            if registry_proc:
+                self._base_layer = registry_proc.add_process_layer()
+        except ValueError:
+            vollog.log(
+                constants.LOGLEVEL_VVVV,
+                "Error walking process list, results may not be valid.",
             )
-            if proc_name == "Registry" and proc.InheritedFromUniqueProcessId == 4:
-                proc_layer_name = proc.add_process_layer()
-                self._base_layer = proc_layer_name
-                break
 
         self._base_block = self.hive.BaseBlock.dereference()
 
@@ -95,6 +94,41 @@ class RegistryHive(linear.LinearlyMappedLayer):
                 constants.LOGLEVEL_VVVV,
                 f"Exception when setting hive {self.name} max address, using {hex(self._maxaddr)}",
             )
+
+    def _find_registry_process(self) -> Optional["extensions.EPROCESS"]:
+        """Walk the active process list and return the Registry process if it exists. Duplicates
+        PsList.list_processes() since pulling in the plugin causes problems.
+
+        Returns:
+            The Registry EPROCESS object if it exists, or None
+        """
+
+        kernel = self.context.modules.get("kernel")
+
+        if not kernel or not kernel.offset:
+            raise ValueError(
+                "Intel layer does not have an associated kernel virtual offset, failing"
+            )
+
+        ps_aph_offset = kernel.get_symbol("PsActiveProcessHead").address
+        list_entry = kernel.object(object_type="_LIST_ENTRY", offset=ps_aph_offset)
+        reloff = kernel.get_type("_EPROCESS").relative_child_offset(
+            "ActiveProcessLinks"
+        )
+        eproc = kernel.object(
+            object_type="_EPROCESS",
+            offset=list_entry.vol.offset - reloff,
+            absolute=True,
+        )
+
+        for proc in eproc.ActiveProcessLinks:
+            proc_name = proc.ImageFileName.cast(
+                "string", max_length=proc.ImageFileName.vol.count, errors="replace"
+            )
+            if proc_name == "Registry" and proc.InheritedFromUniqueProcessId == 4:
+                return proc
+
+        return None
 
     def _get_hive_maxaddr(self, volatile):
         return (

--- a/volatility3/framework/layers/registry.py
+++ b/volatility3/framework/layers/registry.py
@@ -103,7 +103,7 @@ class RegistryHive(linear.LinearlyMappedLayer):
             The Registry EPROCESS object if it exists, or None
         """
 
-        kernel = self.context.modules.get("kernel")
+        kernel = self.context.modules.get(self.config["kernel_module_name"])
 
         if not kernel or not kernel.offset:
             raise ValueError(

--- a/volatility3/framework/layers/registry.py
+++ b/volatility3/framework/layers/registry.py
@@ -95,7 +95,7 @@ class RegistryHive(linear.LinearlyMappedLayer):
                 f"Exception when setting hive {self.name} max address, using {hex(self._maxaddr)}",
             )
 
-    def _find_registry_process(self) -> Optional["extensions.EPROCESS"]:
+    def _find_registry_process(self) -> Optional[extensions.EPROCESS]:
         """Walk the active process list and return the Registry process if it exists. Duplicates
         PsList.list_processes() since pulling in the plugin causes problems.
 

